File: connectors/peertube-ingest/app/minio_client.py
"""MinIO/S3 client.

Uses `boto3` (synchronous) rather than `aioboto3`: S3 calls here are infrequent
relative to the connector's overall throughput -- one `ListObjectsV2`/tagging/
GetObject sequence per ingested recording, and recordings are large video files
where the upload to PeerTube (not the MinIO read) dominates latency. `aioboto3`
also pins to a specific `aiobotocore`/`botocore` combination, which adds
maintenance overhead disproportionate to the benefit here. Every blocking boto3
call is wrapped in `asyncio.to_thread` so the FastAPI event loop (webhook mode)
is never blocked; the batch script runs boto3 directly since it has no event
loop to protect.
"""

from __future__ import annotations

import asyncio
import os
from datetime import datetime, timezone
from typing import Dict, List, Optional

import boto3

from .types import IngestCandidate

MINIO_ENDPOINT = os.environ.get("MINIO_ENDPOINT", "https://minio.example.org")
MINIO_ACCESS_KEY = os.environ.get("MINIO_ACCESS_KEY", "")
MINIO_SECRET_KEY = os.environ.get("MINIO_SECRET_KEY", "")
MINIO_BUCKET = os.environ.get("MINIO_RECORDINGS_BUCKET", "visio-recordings")
MINIO_REGION = os.environ.get("MINIO_REGION", "us-east-1")

_client = None


def _get_client():
    global _client
    if _client is None:
        _client = boto3.client(
            "s3",
            endpoint_url=MINIO_ENDPOINT,
            region_name=MINIO_REGION,
            aws_access_key_id=MINIO_ACCESS_KEY,
            aws_secret_access_key=MINIO_SECRET_KEY,
            # required for MinIO (study 1.3/2.12: self-hosted MinIO, not AWS S3)
            config=boto3.session.Config(s3={"addressing_style": "path"}),
        )
    return _client


def _parse_iso(value: str) -> datetime:
    dt = datetime.fromisoformat(value.replace("Z", "+00:00"))
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    return dt


def list_recent_objects_sync(since_iso: str) -> List[IngestCandidate]:
    """Lists recent objects in the MinIO bucket for batch mode (study 2.12 line 589:
    "this upload can be done as a periodic task -- daily batch"). `since_iso` filters
    client-side on `LastModified` (the S3 API does not offer a server-side date filter).
    """
    since = _parse_iso(since_iso)
    s3 = _get_client()
    candidates: List[IngestCandidate] = []
    continuation_token: Optional[str] = None

    while True:
        kwargs = {"Bucket": MINIO_BUCKET}
        if continuation_token:
            kwargs["ContinuationToken"] = continuation_token
        page = s3.list_objects_v2(**kwargs)
        for obj in page.get("Contents", []):
            last_modified = obj.get("LastModified")
            if obj.get("Key") and last_modified and last_modified >= since:
                candidates.append(
                    IngestCandidate(
                        bucket=MINIO_BUCKET,
                        key=obj["Key"],
                        size=obj.get("Size", 0),
                        last_modified=last_modified.isoformat(),
                    )
                )
        if page.get("IsTruncated"):
            continuation_token = page.get("NextContinuationToken")
        else:
            break

    return candidates


async def list_recent_objects(since_iso: str) -> List[IngestCandidate]:
    return await asyncio.to_thread(list_recent_objects_sync, since_iso)


def get_object_stream_sync(bucket: str, key: str) -> bytes:
    s3 = _get_client()
    result = s3.get_object(Bucket=bucket, Key=key)
    return result["Body"].read()


async def get_object_stream(bucket: str, key: str) -> bytes:
    return await asyncio.to_thread(get_object_stream_sync, bucket, key)


def get_object_tags_sync(bucket: str, key: str) -> Dict[str, str]:
    """Retrieves the object's S3 tags, used in addition to the file name (see
    metadata.py)."""
    s3 = _get_client()
    result = s3.get_object_tagging(Bucket=bucket, Key=key)
    return {tag["Key"]: tag.get("Value", "") for tag in result.get("TagSet", [])}


async def get_object_tags(bucket: str, key: str) -> Dict[str, str]:
    return await asyncio.to_thread(get_object_tags_sync, bucket, key)

File: connectors/peertube-ingest/src/minio-client.ts
import type { Readable } from "node:stream";
import { GetObjectCommand, GetObjectTaggingCommand, ListObjectsV2Command, S3Client } from "@aws-sdk/client-s3";
import { IngestCandidate } from "./types";

const MINIO_ENDPOINT = process.env.MINIO_ENDPOINT ?? "https://minio.example.org";
const MINIO_ACCESS_KEY = process.env.MINIO_ACCESS_KEY ?? "";
const MINIO_SECRET_KEY = process.env.MINIO_SECRET_KEY ?? "";
const MINIO_BUCKET = process.env.MINIO_RECORDINGS_BUCKET ?? "visio-recordings";
const MINIO_REGION = process.env.MINIO_REGION ?? "us-east-1";

let client: S3Client | null = null;

function getClient(): S3Client {
  if (!client) {
    client = new S3Client({
      endpoint: MINIO_ENDPOINT,
      region: MINIO_REGION,
      forcePathStyle: true, // required for MinIO (study 1.3/2.12: self-hosted MinIO, not AWS S3)
      credentials: { accessKeyId: MINIO_ACCESS_KEY, secretAccessKey: MINIO_SECRET_KEY },
    });
  }
  return client;
}

/**
 * Liste les objets recents du bucket MinIO pour le mode batch (etude 2.12 ligne 589 :
 * "ce depot peut se faire en tache periodique -- batch quotidien"). `sinceIso` filtre
 * cote client sur `LastModified` (l'API S3 ne propose pas de filtre serveur par date).
 */
export async function listRecentObjects(sinceIso: string): Promise<IngestCandidate[]> {
  const since = new Date(sinceIso).getTime();
  const s3 = getClient();
  const candidates: IngestCandidate[] = [];
  let continuationToken: string | undefined;

  do {
    const page = await s3.send(
      new ListObjectsV2Command({
        Bucket: MINIO_BUCKET,
        ContinuationToken: continuationToken,
      })
    );
    for (const obj of page.Contents ?? []) {
      if (obj.Key && obj.LastModified && obj.LastModified.getTime() >= since) {
        candidates.push({
          bucket: MINIO_BUCKET,
          key: obj.Key,
          size: obj.Size ?? 0,
          lastModified: obj.LastModified.toISOString(),
        });
      }
    }
    continuationToken = page.NextContinuationToken;
  } while (continuationToken);

  return candidates;
}

export async function getObjectStream(bucket: string, key: string): Promise<Readable> {
  const s3 = getClient();
  const result = await s3.send(new GetObjectCommand({ Bucket: bucket, Key: key }));
  return result.Body as Readable;
}

/** Recupere les tags S3 de l'objet, utilises en complement du nom de fichier (cf. metadata.ts). */
export async function getObjectTags(bucket: string, key: string): Promise<Record<string, string>> {
  const s3 = getClient();
  const result = await s3.send(new GetObjectTaggingCommand({ Bucket: bucket, Key: key }));
  const tags: Record<string, string> = {};
  for (const tag of result.TagSet ?? []) {
    if (tag.Key) tags[tag.Key] = tag.Value ?? "";
  }
  return tags;
}
